File: backend/app/routers/market_data.py
"""
Market Data API Routes

Endpoints for dividends, earnings, recommendations, financials, etc.
"""

from datetime import date
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services.yfinance_service import YFinanceService
from app.services.finnhub_service import FinnhubService

router = APIRouter(prefix="/market-data", tags=["Market Data"])

# Service instances
yfinance_service = YFinanceService()
finnhub_service = FinnhubService()


# ============ yfinance endpoints (no rate limit) ============

@router.get("/dividends/{ticker}")
async def get_dividends(
    ticker: str,
    refresh: bool = Query(False, description="Force refresh from API"),
    db: AsyncSession = Depends(get_db),
):
    """
    Get dividend history for a stock.

    Source: yfinance (no rate limit)
    Cache: 24 hours
    """
    dividends = await yfinance_service.get_dividends(db, ticker, force_refresh=refresh)
    return {"ticker": ticker.upper(), "dividends": dividends}


@router.get("/earnings-calendar/{ticker}")
async def get_earnings_calendar(
    ticker: str,
    refresh: bool = Query(False),
    db: AsyncSession = Depends(get_db),
):
    """
    Get next earnings date and estimates.

    Source: yfinance (no rate limit)
    Cache: 6 hours
    """
    data = await yfinance_service.get_earnings_calendar(db, ticker, force_refresh=refresh)
    return {"ticker": ticker.upper(), **data}


@router.get("/recommendations/{ticker}")
async def get_recommendations(
    ticker: str,
    refresh: bool = Query(False),
    db: AsyncSession = Depends(get_db),
):
    """
    Get analyst recommendations (buy/hold/sell).

    Source: yfinance (no rate limit)
    Cache: 6 hours
    """
    recs = await yfinance_service.get_analyst_recommendations(db, ticker, force_refresh=refresh)

    # Calculate summary
    if recs:
        latest = recs[0]
        total = latest['strong_buy'] + latest['buy'] + latest['hold'] + latest['sell'] + latest['strong_sell']
        summary = {
            "total_analysts": total,
            "buy_percent": round((latest['strong_buy'] + latest['buy']) / total * 100, 1) if total else 0,
            "hold_percent": round(latest['hold'] / total * 100, 1) if total else 0,
            "sell_percent": round((latest['sell'] + latest['strong_sell']) / total * 100, 1) if total else 0,
        }
    else:
        summary = {}

    return {"ticker": ticker.upper(), "summary": summary, "history": recs}


@router.get("/financials/{ticker}")
async def get_financials(
    ticker: str,
    statement: str = Query("income", enum=["income", "balance", "cashflow"]),
    period: str = Query("quarterly", enum=["quarterly", "annual"]),
    refresh: bool = Query(False),
    db: AsyncSession = Depends(get_db),
):
    """
    Get financial statements.

    Source: yfinance (no rate limit)
    Cache: 24 hours
    """
    data = await yfinance_service.get_financial_statements(
        db, ticker,
        statement_type=statement,
        period=period,
        force_refresh=refresh
    )
    return {
        "ticker": ticker.upper(),
        "statement_type": statement,
        "period": period,
        "statements": data,
    }


# ============ Finnhub endpoints (60 calls/min limit) ============

@router.get("/price-target/{ticker}")
async def get_price_target(
    ticker: str,
    refresh: bool = Query(False),
    db: AsyncSession = Depends(get_db),
):
    """
    Get analyst price targets.

    Source: Finnhub (rate limited)
    Cache: 6 hours
    """
    data = await finnhub_service.get_price_target(db, ticker, force_refresh=refresh)
    if not data:
        raise HTTPException(404, f"No price target data for {ticker}")
    return {"ticker": ticker.upper(), **data}


@router.get("/earnings-surprises/{ticker}")
async def get_earnings_surprises(
    ticker: str,
    limit: int = Query(4, ge=1, le=20),
    refresh: bool = Query(False),
    db: AsyncSession = Depends(get_db),
):
    """
    Get earnings surprises (actual vs estimate).

    Source: Finnhub (rate limited)
    Cache: 6 hours
    """
    data = await finnhub_service.get_earnings_surprises(db, ticker, limit, force_refresh=refresh)
    return {"ticker": ticker.upper(), "earnings": data}


@router.get("/ipo-calendar")
async def get_ipo_calendar(
    from_date: Optional[date] = Query(None, description="Start date (default: today)"),
    to_date: Optional[date] = Query(None, description="End date (default: +30 days)"),
    refresh: bool = Query(False),
):
    """
    Get upcoming IPOs.

    Source: Finnhub (rate limited)
    Cache: 1 hour
    """
    data = await finnhub_service.get_ipo_calendar(from_date, to_date, force_refresh=refresh)
    return {"ipos": data}


@router.get("/economic-calendar")
async def get_economic_calendar(
    from_date: Optional[date] = Query(None, description="Start date (default: today)"),
    to_date: Optional[date] = Query(None, description="End date (default: +14 days)"),
    refresh: bool = Query(False),
):
    """
    Get economic events (Fed meetings, CPI, jobs, etc.).

    Source: Finnhub (rate limited)
    Cache: 1 hour
    """
    data = await finnhub_service.get_economic_calendar(from_date, to_date, force_refresh=refresh)
    return {"events": data}


# ============ Combined endpoint for stock detail page ============

@router.get("/stock-overview/{ticker}")
async def get_stock_overview(
    ticker: str,
    db: AsyncSession = Depends(get_db),
):
    """
    Get all market data for a stock in one call.

    Combines: earnings calendar, recommendations, price target.
    Useful for stock detail pages.
    """
    # These run in parallel (sort of - Python async)
    earnings = await yfinance_service.get_earnings_calendar(db, ticker)
    recommendations = await yfinance_service.get_analyst_recommendations(db, ticker)
    price_target = await finnhub_service.get_price_target(db, ticker)

    # Summary
    rec_summary = {}
    if recommendations:
        latest = recommendations[0]
        total = latest['strong_buy'] + latest['buy'] + latest['hold'] + latest['sell'] + latest['strong_sell']
        if total > 0:
            buy_count = latest['strong_buy'] + latest['buy']
            sell_count = latest['sell'] + latest['strong_sell']
            rec_summary = {
                "consensus": "BUY" if buy_count > total / 2 else (
                    "SELL" if sell_count > total / 2 else "HOLD"
                ),
                "total_analysts": total,
            }

    return {
        "ticker": ticker.upper(),
        "earnings": earnings,
        "recommendations": rec_summary,
        "price_target": price_target,
    }

File: frontend/src/api/marketData.ts
/**
 * Market Data API Client
 *
 * Provides access to market data from yfinance and Finnhub:
 * - Dividends, earnings, recommendations, financials (yfinance)
 * - Price targets, earnings surprises, IPO calendar, economic calendar (Finnhub)
 */

import apiClient from './client';

const MARKET_DATA_BASE = '/api/v1/market-data';

export interface Dividend {
  ex_date: string;
  amount: number;
}

export interface EarningsCalendar {
  earnings_date?: string;
  eps_estimate?: number;
  revenue_estimate?: number;
}

export interface RecommendationPeriod {
  period: string;
  strong_buy: number;
  buy: number;
  hold: number;
  sell: number;
  strong_sell: number;
}

export interface RecommendationsResponse {
  ticker: string;
  summary: {
    total_analysts: number;
    buy_percent: number;
    hold_percent: number;
    sell_percent: number;
  };
  history: RecommendationPeriod[];
}

export interface FinancialStatement {
  period_end: string;
  data: Record<string, number | string>;
}

export interface PriceTarget {
  target_high?: number;
  target_low?: number;
  target_mean?: number;
  target_median?: number;
  last_updated?: string;
}

export interface EarningsSurprise {
  period: string;
  actual: number;
  estimate: number;
  surprise: number;
  surprise_percent: number;
}

export interface IPOCalendarItem {
  symbol?: string;
  company_name: string;
  exchange?: string;
  ipo_date?: string;
  price_range?: string;
  shares?: number;
  status?: string;
}

export interface EconomicCalendarEvent {
  event: string;
  country: string;
  date?: string;
  time?: string;
  impact?: string;
  actual?: string;
  estimate?: string;
  previous?: string;
  unit?: string;
}

export interface StockOverview {
  ticker: string;
  earnings: EarningsCalendar;
  recommendations: {
    consensus?: string;
    total_analysts?: number;
  };
  price_target?: PriceTarget;
}

export const marketDataApi = {
  /**
   * Get dividend history for a stock (yfinance - no rate limit)
   */
  getDividends: async (ticker: string, refresh: boolean = false): Promise<{ ticker: string; dividends: Dividend[] }> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/dividends/${ticker}`, {
      params: { refresh },
    });
    return response.data;
  },

  /**
   * Get earnings calendar (next earnings date, estimates) (yfinance - no rate limit)
   */
  getEarningsCalendar: async (ticker: string, refresh: boolean = false): Promise<{ ticker: string } & EarningsCalendar> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/earnings-calendar/${ticker}`, {
      params: { refresh },
    });
    return response.data;
  },

  /**
   * Get analyst recommendations (buy/hold/sell) (yfinance - no rate limit)
   */
  getRecommendations: async (ticker: string, refresh: boolean = false): Promise<RecommendationsResponse> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/recommendations/${ticker}`, {
      params: { refresh },
    });
    return response.data;
  },

  /**
   * Get financial statements (income, balance, cashflow) (yfinance - no rate limit)
   */
  getFinancials: async (
    ticker: string,
    statement: 'income' | 'balance' | 'cashflow' = 'income',
    period: 'quarterly' | 'annual' = 'quarterly',
    refresh: boolean = false
  ): Promise<{ ticker: string; statement_type: string; period: string; statements: FinancialStatement[] }> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/financials/${ticker}`, {
      params: { statement, period, refresh },
    });
    return response.data;
  },

  /**
   * Get analyst price targets (Finnhub - rate limited, cached)
   */
  getPriceTarget: async (ticker: string, refresh: boolean = false): Promise<{ ticker: string } & PriceTarget> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/price-target/${ticker}`, {
      params: { refresh },
    });
    return response.data;
  },

  /**
   * Get earnings surprises (actual vs estimate) (Finnhub - rate limited, cached)
   */
  getEarningsSurprises: async (
    ticker: string,
    limit: number = 4,
    refresh: boolean = false
  ): Promise<{ ticker: string; earnings: EarningsSurprise[] }> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/earnings-surprises/${ticker}`, {
      params: { limit, refresh },
    });
    return response.data;
  },

  /**
   * Get IPO calendar (Finnhub - rate limited, cached)
   */
  getIpoCalendar: async (
    fromDate?: string,
    toDate?: string,
    refresh: boolean = false
  ): Promise<{ ipos: IPOCalendarItem[] }> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/ipo-calendar`, {
      params: { from_date: fromDate, to_date: toDate, refresh },
    });
    return response.data;
  },

  /**
   * Get economic calendar (Fed meetings, CPI, jobs, etc.) (Finnhub - rate limited, cached)
   */
  getEconomicCalendar: async (
    fromDate?: string,
    toDate?: string,
    refresh: boolean = false
  ): Promise<{ events: EconomicCalendarEvent[] }> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/economic-calendar`, {
      params: { from_date: fromDate, to_date: toDate, refresh },
    });
    return response.data;
  },

  /**
   * Get all market data for a stock in one call (combined endpoint)
   */
  getStockOverview: async (ticker: string): Promise<StockOverview> => {
    const response = await apiClient.get(`${MARKET_DATA_BASE}/stock-overview/${ticker}`);
    return response.data;
  },
};

